cli/cmds/cluster_delete.go

```diff
@@ -55,6 +65,52 @@ func delete(clx *cli.Context) error {
 			Namespace: Namespace(),
 		},
 	}
+	// keep bootstrap secrets and tokens if --keep-data flag is passed
+	if keepData {
+		// skip removing tokenSecret
+		if err := RemoveOwnerReferenceFromSecret(ctx, k3kcluster.TokenSecretName(cluster.Name), ctrlClient, cluster); err != nil {
+			return err
+		}
+
+		// skip removing webhook secret
+		if err := RemoveOwnerReferenceFromSecret(ctx, agent.WebhookSecretName(cluster.Name), ctrlClient, cluster); err != nil {
+			return err
+		}
+	}
+
+	if err := ctrlClient.Delete(ctx, &cluster); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	// make sure to delete pv claims for the cluster if --keep-data is not used
+	if !keepData {
+		matchingLabels := client.MatchingLabels(map[string]string{"cluster": cluster.Name, "role": "server"})
+		listOpts := client.ListOptions{Namespace: cluster.Namespace}
+		matchingLabels.ApplyToList(&listOpts)
+		deleteOpts := &client.DeleteAllOfOptions{ListOptions: listOpts}
+
+		if err := ctrlClient.DeleteAllOf(ctx, &v1.PersistentVolumeClaim{}, deleteOpts); err != nil {
+			return client.IgnoreNotFound(err)
+		}
+	}
+
+	return nil
+}
+
+func RemoveOwnerReferenceFromSecret(ctx context.Context, name string, cl client.Client, cluster v1alpha1.Cluster) error {
+	var secret v1.Secret
+	key := types.NamespacedName{
+		Name:      name,
+		Namespace: cluster.Namespace,
+	}
+
+	if err := cl.Get(ctx, key, &secret); err != nil {
+		return err
+	}
+
+	if err := controllerutil.RemoveOwnerReference(&cluster, &secret, cl.Scheme()); err != nil {
+		return client.IgnoreNotFound(err)
+	}
```

Comment:
The `RemoveOwnerReference` does not fail with a ErrNotFound, but the `Get` yes. We should do the check there, and log. Also we should check if the object has the owner reference, or it will return an error.

https://github.com/kubernetes-sigs/controller-runtime/blob/v0.20.3/pkg/controller/controllerutil/controllerutil.go#L142